src/batch/src/executor/s3_file_scan.rs

```diff
@@ -0,0 +1,163 @@
+use risingwave_common::array::arrow::IcebergArrowConvert;
+use risingwave_common::catalog::Schema;
+
+use crate::error::BatchError;
+use crate::executor::{DataChunk, Executor};
+
+#[derive(PartialEq, Debug)]
+pub enum FileFormat {
+    Parquet,
+}
+
+/// S3 file scan executor. Currently only support parquet file format.
+pub struct S3FileScanExecutor {
+    file_format: FileFormat,
+    location: String,
+    s3_region: String,
+    s3_access_key: String,
+    s3_secret_key: String,
+    batch_size: usize,
+    schema: Schema,
+    identity: String,
+}
+
+impl Executor for S3FileScanExecutor {
+    fn schema(&self) -> &risingwave_common::catalog::Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> super::BoxedDataChunkStream {
+        self.do_execute().boxed()
+    }
+}
+
+impl S3FileScanExecutor {
+    #![expect(dead_code)]
+    pub fn new(
+        file_format: FileFormat,
+        location: String,
+        s3_region: String,
+        s3_access_key: String,
+        s3_secret_key: String,
+        batch_size: usize,
+        schema: Schema,
+        identity: String,
+    ) -> Self {
+        Self {
+            file_format,
+            location,
+            s3_region,
+            s3_access_key,
+            s3_secret_key,
+            batch_size,
+            schema,
+            identity,
+        }
+    }
+
+    #[try_stream(ok = DataChunk, error = BatchError)]
+    async fn do_execute(self: Box<Self>) {
+        assert_eq!(self.file_format, FileFormat::Parquet);
+
+        let mut props = HashMap::new();
+        props.insert(S3_REGION, self.s3_region.clone());
+        props.insert(S3_ACCESS_KEY_ID, self.s3_access_key.clone());
+        props.insert(S3_SECRET_ACCESS_KEY, self.s3_secret_key.clone());
+
+        let file_io_builder = FileIOBuilder::new("s3");
+        let file_io = file_io_builder.with_props(props.into_iter()).build()?;
+        let parquet_file = file_io.new_input(&self.location)?;
+
+        let parquet_metadata = parquet_file.metadata().await?;
+        let parquet_reader = parquet_file.reader().await?;
+        let arrow_file_reader = ArrowFileReader::new(parquet_metadata, parquet_reader);
+
+        let mut batch_stream_builder = ParquetRecordBatchStreamBuilder::new(arrow_file_reader)
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        let arrow_schema = batch_stream_builder.schema();
+        assert_eq!(arrow_schema.fields.len(), self.schema.fields.len());
+        for (field, arrow_field) in self.schema.fields.iter().zip(arrow_schema.fields.iter()) {
+            assert_eq!(*field.name, *arrow_field.name());
+        }
+
+        batch_stream_builder = batch_stream_builder.with_projection(ProjectionMask::all());
+
+        batch_stream_builder = batch_stream_builder.with_batch_size(self.batch_size);
+
+        let record_batch_stream = batch_stream_builder.build().map_err(|e| anyhow!(e))?;
```

Comment:
Recommend using some specific error instead of `anyhow!(e)`

Reply:
Sure, I will improve it together with the next PR.